src/librustc/ty/util.rs

```diff
@@ -700,6 +700,83 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
     }
 
     #[inline]
+    pub fn may_drop(&'tcx self, tcx: TyCtxt<'a, 'tcx, 'tcx>) -> bool {
+        if self.flags.get().intersects(TypeFlags::MAY_DROP_CACHED) {
+            return self.flags.get().intersects(TypeFlags::MAY_DROP);
+        }
+
+        self.may_drop_inner(tcx, &mut FxHashSet())
+    }
+
+    fn may_drop_inner(&'tcx self,
+                      tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                      visited: &mut FxHashSet<Ty<'tcx>>)
+                      -> bool {
+        if self.flags.get().intersects(TypeFlags::MAY_DROP_CACHED) {
+            return self.flags.get().intersects(TypeFlags::MAY_DROP);
+        }
+
+        // This should be reported as an error by `check_representable`.
+        //
+        // Consider the type as not needing drop in the meanwhile to avoid
+        // further errors.
+        if visited.replace(self).is_some() {
```

Comment:
not that I care very much...but why `replace(self).is_some()` and `not `insert(self)`?

Also, I think it's worth expanding the comment to clarify that this condition indicates a cycle -- specifically because there is no `CACHED` flag set. At first I was confused because a "cross-edge" in the graph (i.e., visiting something you already completely visited before) seemed like it would also be true here.